File: apps/mobile/modules/import-sources/ios/StreamCopier.swift
import Foundation

/// File copy for `copyToPath`: one read produces the copy, the SHA-256, and
/// a first-chunk mime sniff. Writes `destPath` directly. The caller supplies
/// the claim-scoped temp path and owns atomic publication; a native-internal
/// `.tmp` suffix would corrupt the orphan scanner's filename parsing. On any
/// failure or cancellation the partial dest is deleted before the error
/// propagates.
public enum StreamCopier {
  public struct CopyResult: Equatable {
    public let size: Int64
    public let sha256Hex: String
    public let mime: String?

    public init(size: Int64, sha256Hex: String, mime: String?) {
      self.size = size
      self.sha256Hex = sha256Hex
      self.mime = mime
    }
  }

  public static func copy(
    sourcePath: String,
    destPath: String,
    copyId: String? = nil,
    registry: CopyRegistry = .shared,
    chunkSize: Int = 65536,
    // Injectable for fail-mid-stream tests; production writes the dest handle.
    writeOverride: ((Data) throws -> Void)? = nil,
    onBytes: ((Int64) -> Void)? = nil
  ) throws -> CopyResult {
    let fm = FileManager.default

    guard fm.fileExists(atPath: sourcePath) else {
      throw CodedError("deleted", "source missing: \(sourcePath)")
    }
    let input: FileHandle
    do {
      input = try FileHandle(forReadingFrom: URL(fileURLWithPath: sourcePath))
    } catch {
      throw mapReadOpenError(error)
    }
    defer { try? input.close() }

    fm.createFile(atPath: destPath, contents: nil)
    let output: FileHandle?
    if writeOverride == nil {
      guard let handle = FileHandle(forWritingAtPath: destPath) else {
        try? fm.removeItem(atPath: destPath)
        throw CodedError("io-error", "cannot open dest: \(destPath)")
      }
      output = handle
    } else {
      output = nil
    }

    var sink = Sha256Sink()
    var size: Int64 = 0
    var mime: String?

    func cleanupAndThrow(_ error: Error) throws -> Never {
      try? output?.close()
      try? fm.removeItem(atPath: destPath)
      throw error
    }

    while true {
      if let copyId, registry.isCancelled(copyId) {
        try cleanupAndThrow(CodedError("cancelled", "copy cancelled"))
      }
      let chunk = input.readData(ofLength: chunkSize)
      if chunk.isEmpty { break }
      if size == 0 { mime = MimeSniffer.sniff(chunk) }
      do {
        if let writeOverride {
          try writeOverride(chunk)
        } else {
          try output?.write(contentsOf: chunk)
        }
      } catch {
        try cleanupAndThrow(mapWriteError(error))
      }
      sink.update(chunk)
      size += Int64(chunk.count)
      onBytes?(size)
    }

    if let copyId, registry.isCancelled(copyId) {
      try cleanupAndThrow(CodedError("cancelled", "copy cancelled"))
    }
    try? output?.close()
    return CopyResult(size: size, sha256Hex: sink.finalizeHex(), mime: mime)
  }

  private static func mapReadOpenError(_ error: Error) -> Error {
    if let coded = error as? CodedError { return coded }
    let ns = error as NSError
    if ns.domain == NSCocoaErrorDomain {
      if ns.code == NSFileReadNoSuchFileError { return CodedError("deleted", ns.localizedDescription) }
      if ns.code == NSFileReadNoPermissionError {
        return CodedError("permission-denied", ns.localizedDescription)
      }
    }
    return CodedError("io-error", ns.localizedDescription)
  }

  private static func mapWriteError(_ error: Error) -> Error {
    if let coded = error as? CodedError { return coded }
    let ns = error as NSError
    if ns.domain == NSCocoaErrorDomain, ns.code == NSFileWriteOutOfSpaceError {
      return CodedError("not-enough-space", ns.localizedDescription)
    }
    if ns.domain == NSPOSIXErrorDomain, ns.code == Int(ENOSPC) {
      return CodedError("not-enough-space", ns.localizedDescription)
    }
    return CodedError("io-error", ns.localizedDescription)
  }
}

File: apps/mobile/modules/import-sources/android/src/main/java/expo/modules/importsources/StreamCopier.kt
package expo.modules.importsources

import android.content.ContentResolver
import android.net.Uri
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStream

/**
 * Copy for `copyToPath` that hashes in the same read: one pass produces both
 * the bytes on disk and the SHA-256. Writes destPath directly; the caller
 * supplies the claim-scoped temp path and owns atomic publication. The
 * partial dest is deleted on any failure or cancellation. `content://`
 * sources stream through the ContentResolver, never a filesystem-path
 * assumption.
 */
object StreamCopier {
  // headBytes: the file's first bytes (up to 32), captured across reads. Type
  // classification is the JS side's job; native only saves it a read.
  data class CopyResult(val size: Long, val sha256Hex: String, val headBytes: ByteArray)

  const val HEAD_BYTE_COUNT = 32

  fun copy(
    resolver: ContentResolver?,
    srcUri: String,
    destPath: String,
    copyId: String? = null,
    registry: CopyRegistry = CopyRegistry.shared,
    chunkSize: Int = 65536,
    // Injectable for fail-mid-stream tests; production writes the dest stream.
    writeOverride: ((ByteArray, Int) -> Unit)? = null,
    onBytes: ((Long) -> Unit)? = null,
  ): CopyResult {
    val input: InputStream
    try {
      if (srcUri.startsWith("content://")) {
        val resolverRequired =
          resolver ?: throw CodedError("io-error", "no resolver for content uri")
        input =
          resolverRequired.openInputStream(Uri.parse(srcUri))
            ?: throw CodedError("deleted", "provider returned no stream: $srcUri")
      } else {
        input = FileInputStream(SourceRefCodec.pathFromFileUriOrPath(srcUri))
      }
    } catch (e: CodedError) {
      throw e
    } catch (e: FileNotFoundException) {
      throw CodedError("deleted", e.message ?: srcUri)
    } catch (e: SecurityException) {
      throw CodedError("permission-denied", e.message ?: srcUri)
    } catch (e: IOException) {
      throw CodedError("io-error", e.message ?: srcUri)
    }

    val dest = File(SourceRefCodec.pathFromFileUriOrPath(destPath))
    val output = if (writeOverride == null) dest.outputStream() else null
    val sink = Sha256Sink()
    var size = 0L
    val buffer = ByteArray(chunkSize)
    // The head window accumulates across reads: a content:// stream may
    // legally return only a few bytes on its first read.
    val head = ByteArray(HEAD_BYTE_COUNT)
    var headLen = 0

    fun cleanupAndThrow(error: Throwable): Nothing {
      try {
        output?.close()
      } catch (_: IOException) {}
      dest.delete()
      throw error
    }

    input.use { source ->
      while (true) {
        if (copyId != null && registry.isCancelled(copyId)) {
          cleanupAndThrow(CodedError("cancelled", "copy cancelled"))
        }
        val read =
          try {
            source.read(buffer)
          } catch (e: IOException) {
            // A stream closed by cancellation is a cancel, not an io failure.
            if (copyId != null && registry.isCancelled(copyId)) {
              cleanupAndThrow(CodedError("cancelled", "copy cancelled"))
            }
            cleanupAndThrow(mapIoError(e))
          }
        if (read < 0) break
        try {
          if (writeOverride != null) {
            writeOverride(buffer, read)
          } else {
            output?.write(buffer, 0, read)
          }
        } catch (e: Exception) {
          cleanupAndThrow(mapIoError(e))
        }
        sink.update(buffer, read)
        if (headLen < head.size && read > 0) {
          val n = minOf(read, head.size - headLen)
          System.arraycopy(buffer, 0, head, headLen, n)
          headLen += n
        }
        size += read
        if (copyId == null || !registry.isCancelled(copyId)) {
          onBytes?.invoke(size)
        }
      }
    }

    if (copyId != null && registry.isCancelled(copyId)) {
      cleanupAndThrow(CodedError("cancelled", "copy cancelled"))
    }
    try {
      // close() flushes, so ENOSPC can surface here; it must leave neither a
      // partial dest nor an un-coded exception.
      output?.close()
    } catch (e: IOException) {
      cleanupAndThrow(mapIoError(e))
    }
    return CopyResult(size, sink.finalizeHex(), head.copyOf(headLen))
  }

}
